aws-lc-rs/src/kem.rs

```diff
@@ -0,0 +1,504 @@
+        })
+    }
+}
+
+unsafe impl<Id> Send for EncapsulationKey<Id> where Id: AlgorithmIdentifier {}
+
+unsafe impl<Id> Sync for EncapsulationKey<Id> where Id: AlgorithmIdentifier {}
+
+impl<Id> AsDer<EncapsulationKeyBinDer> for EncapsulationKey<Id>
+where
+    Id: AlgorithmIdentifier,
+{
+    fn as_der(&self) -> Result<EncapsulationKeyBinDer, Unspecified> {
+        let mut public_key_size = self.algorithm.encapsulate_key_size();
+        let mut pubkey_bytes = vec![0u8; public_key_size];
+        if 1 != unsafe {
+            EVP_PKEY_get_raw_public_key(
+                self.evp_pkey.as_const_ptr(),
+                pubkey_bytes.as_mut_ptr(),
+                &mut public_key_size,
+            )
+        } {
+            return Err(Unspecified);
+        }
+
+        pubkey_bytes.truncate(public_key_size);
+
+        Ok(Buffer::new(pubkey_bytes))
+    }
+}
+
+impl<Id> Debug for EncapsulationKey<Id>
+where
+    Id: AlgorithmIdentifier,
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("EncapsulationKey")
+            .field("algorithm", &self.algorithm)
+            .finish_non_exhaustive()
+    }
+}
+
+/// A set of encrypted bytes produced by [`EncapsulationKey::encapsulate`], and used as an input to [`DecapsulationKey::decapsulate`].
+pub struct Ciphertext<'a>(Cow<'a, [u8]>);
+
+impl<'a> Ciphertext<'a> {
+    fn new(value: Vec<u8>) -> Ciphertext<'a> {
+        Self(Cow::Owned(value))
+    }
+}
+
+impl<'a> Drop for Ciphertext<'a> {
+    fn drop(&mut self) {
+        if let Cow::Owned(ref mut v) = self.0 {
+            v.zeroize();
+        }
+    }
+}
+
+impl<'a> AsRef<[u8]> for Ciphertext<'a> {
+    fn as_ref(&self) -> &[u8] {
+        match self.0 {
+            Cow::Borrowed(v) => v,
+            Cow::Owned(ref v) => v.as_ref(),
+        }
+    }
+}
+
+impl<'a> From<&'a [u8]> for Ciphertext<'a> {
+    fn from(value: &'a [u8]) -> Self {
+        Self(Cow::Borrowed(value))
+    }
+}
+
+/// The cryptograpic shared secret output from the KEM encapsulate / decapsulate process.
+pub struct SharedSecret(Box<[u8]>);
+
+impl SharedSecret {
+    fn new(value: Box<[u8]>) -> Self {
+        Self(value)
+    }
+}
+
+impl Drop for SharedSecret {
+    fn drop(&mut self) {
+        self.0.zeroize();
+    }
+}
+
+impl AsRef<[u8]> for SharedSecret {
+    fn as_ref(&self) -> &[u8] {
+        self.0.as_ref()
+    }
+}
+
+// Returns an LcPtr to an EVP_PKEY
+#[inline]
+fn kem_key_generate(nid: i32) -> Result<LcPtr<EVP_PKEY>, Unspecified> {
+    let ctx = LcPtr::new(unsafe { EVP_PKEY_CTX_new_id(EVP_PKEY_KEM, null_mut()) })?;
+    if 1 != unsafe { EVP_PKEY_CTX_kem_set_params(*ctx, nid) }
+        || 1 != unsafe { EVP_PKEY_keygen_init(*ctx) }
+    {
+        return Err(Unspecified);
+    }
+
+    let mut key_raw: *mut EVP_PKEY = null_mut();
+    if 1 != unsafe { EVP_PKEY_keygen(*ctx, &mut key_raw) } {
+        return Err(Unspecified);
+    }
+    Ok(LcPtr::new(key_raw)?)
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
can you please add a comment about what this test does?

Reply:
It's not a test, it is a Rust module, so there is nothing really to document since its a standard idiom.